### PR TITLE
analyzer2.txt

### PR DESCRIPTION
**Pull request by rjmac**

### CODE REVIEW

**Review by rlvoyer — APPROVED** [review]
This is awesome. Thanks for taking the time to write this up. Left some comments. Another pass as time permits to generate some simple diagrams (for demonstrating class relationships or data flow, for example) would also be super valuable.

File: doc/analyzer2.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+Many of the classes in analyzer2 are very generic.  In order to keep
+this noise down to a minimum, the common generic parameters are
+bundled up into a single `MetaTypes` trait, which just contains the
+other types.
+
+# TableFinder
+
+A `TableFinder` is the library's interface to a catalog of tables,
+saved views, and UDFs.  It takes a user's query and finds all the
```
[the remark this answers is not in the document]
Comment (rjmac) [reply doc/analyzer2.md:21]: Nothing; those are not tables and are therefore out of scope for a tablefinder (which doesn't actually know about "views" per se, only a more generic "saved query" concept).

File: doc/analyzer2.md [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+by one of the coarser sets is issued, if all agggregates can be
+expressed in terms of rolling up the finer aggregates (hence the name
+of the feature) the system can use the rollup to accelerate the query.
+
+This rollup is applied in
+`RollupExact#rewriteAggregatedOnAggregatedDifferentGroupBy`
+
+Three interfaces in the rollup machinery can be implemented to perform
+those further rollups in different ways: `SemigroupRewriter`,
+``FunctionSplitter`, and `AdHocRewriter`.  In addition, the
```
Comment (saraid) [inline doc/analyzer2.md:358]:
```suggestion
`FunctionSplitter`, and `AdHocRewriter`.  In addition, the
```

File: doc/analyzer2.md [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+class Something[MT <: MetaTypes] { // no mixin
+    def takesAColumnType(typ: MT#ColumnType)
+}
+
+class Something[MT <: MetaTypes] extends MetaTypeHelper[MT] {
+    def takesAColumnType(typ: CT)
+}
+```
+
+### LabelUniverse
```
[the remark this answers is not in the document]
Comment (rjmac) [reply doc/analyzer2.md:61]: They absolutely should be, but not here.

File: doc/analyzer2.md [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+sub-computations are `sum` and `count` and the recombining function is
+`/`.  This trait expresses this kind of rewrite (and honestly, exists
+purely to support `avg`)
+
+### AdHocRewriter
+
+This is the most general interface.  It turns an expression into a
+collection of alternate equivalent expressions which may be easier to
+roll up.  For example, if you have an expression like
+`some_timestamp() < '2001-10-05T00:00:00`, this is the same as
```
Comment (saraid) [inline doc/analyzer2.md:386]:
```suggestion
`some_timestamp() < '2001-10-05T00:00:00'`, this is the same as
```

File: doc/analyzer2.md [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+`date_trunc_ymd(some_timestamp()) < '2001-10-05T00:00:00` since the
```
Comment (saraid) [inline doc/analyzer2.md:387]:
```suggestion
`date_trunc_ymd(some_timestamp()) < '2001-10-05T00:00:00'` since the
```

**Review by saraid — APPROVED** [review]

**Comment by rjmac** [comment]
comment comment

**Pull request merged**